File: HW3/HW3/SortedArr.cs
﻿using System;

namespace HW3
{
    class SortedArr<Type>
    {
        private Func<object, object, bool> isEqual; 
        private Func<object, object, bool> isGreater;
        private object[] arr;  // our queue
        protected int indexer=0;
        public SortedArr(Func<object, object, bool> isGreater, Func<object, object, bool> isEqual) // constructor
        {
            this.isGreater = isGreater;
            this.isEqual = isEqual;
        }

        public void add(Type v) // add v to the queue
        {
 
            indexer++;
            object []tmp= new object[indexer];
           

            if (indexer == 1)
            {
                tmp[0] = v;
                arr = tmp;
                return;
            }
            bool placed = false;
            int i = 0;
            while (!placed)
            {
               
                if (i<indexer-1 && isGreater((object)v,arr[i]))
                {
                    tmp[i] = arr[i];
                }
                else
                {
                    tmp[i] = v;
                    placed = true;
                }
                i++;
            }
            while (i < indexer)
            {
                tmp[i] = arr[i - 1];
                i++;
            }
        
            arr = tmp;
            
           
        }

        public bool delete(Type v)  // delete v from the queue
        {
            if (indexer <=0)
            {
                return false;
            }
            int found=0;
            for (int i=0 ;i<indexer;i++)
            {
                if (isEqual(arr[i],(object) v)&&(found==0))
                {
                    found = 1;
                }
                else if (found==1)
                {
                    arr[i-1] = arr[i];
                }
            }

            if (found == 0)
                return false;
      

            indexer--;
            object[] tmp = new object[indexer];
            for (int i = 0; i < indexer; i++)
                tmp[i] = arr[i];
            arr = tmp;
            return true;  
          
        }

        public Type Geti(int i) // return the value of index i from the queue
        {
            return (Type)arr[i];
        }

        public int N { get { return indexer; } }
    }
}

File: HW5/HW5/SortedArr.cpp
/*
#include <iostream>
#include "SortedArr.h"
using namespace std;
using namespace HW5;


	template <class type>
	SortedArr<type>::SortedArr(bool(*isGreater)(void *, void *), bool(*isEqual)(void *, void *))// constructor
	{
		this.isGreater = isGreater;
		this.isEqual = isEqual;
	}

	template <class type>
	SortedArr<type>::~SortedArr()// destructor
	{
		for (int i; i < indexer; i++)
		{
			delete arr[i];
		}
		delete[]arr;
	}

	template <class type>
	void SortedArr<type>::add(type v) // add v to the queue
	{

		indexer++;
		(void*)[]tmp = new (void*)[indexer];


		if (indexer == 1)
		{
			tmp[0] = &(new type);
			(type)(*tmp[0]) = v;
			arr = tmp;
			return;
		}
		bool placed = false;
		int i = 0;
		while (!placed)
		{

			if (i < indexer - 1 && isGreater(&v, arr[i]))
			{
				tmp[i] = arr[i];
			}
			else
			{
				tmp[i] = &(new type);
				(type)(*tmp[i]) = v;


				placed = true;
			}
			i++;
		}
		while (i < indexer)
		{
			tmp[i] = arr[i - 1];
			i++;
		}

		arr = tmp;


	}


	template <class type>
	bool SortedArr<type>::delete_member(type v)  // delete v from the queue
	{
		if (indexer <= 0)
		{
			return false;
		}
		int found = 0;
		for (int i = 0; i < indexer; i++)
		{
			if (isEqual(arr[i], &v) && (found == 0))
			{
				found = 1;
				delete arr[i];
			}
			else if (found == 1)
			{
				arr[i - 1] = arr[i];

			}
		}
		 
		if (found == 0)
			return false;
		

		indexer--;
		type ** tmp = new type* [indexer];
		for (int i = 0; i < indexer; i++)
		tmp[i] = arr[i];
		delete []arr;
		arr = tmp;
		return true;

	}




	template <class type>
	type SortedArr<type>::Geti(int i) // return the value of index i from the queue
	{
		return (type)arr[i];
	}

	template <class type>
	int SortedArr<type>::getN() { return indexer; } // getter for size
	*/
